aten/src/ATen/native/SpectralOps.cpp

```diff
@@ -17,23 +17,212 @@
 #include <cmath>
 
 namespace at { namespace native {
+static inline Tensor _fft(
+    const Tensor &self, int64_t signal_ndim, bool complex_input,
+    const bool complex_output, bool inverse, IntArrayRef signal_sizes,
+    fft_norm_mode normalization, bool onesided);
+
+namespace {
+
+ScalarType promote_type(ScalarType type, bool require_complex) {
```

Comment:
`promote_type` -> `promote_to_complex`? Something more precise would be nice.

Reply:
The result can be a real dtype if `require_complex=False` is specified.

Reply:
Good point. "promote_type_fft"? We already have "promote_types" which is a more general function with a very similar name.

Reply:
Renamed to `promote_type_fft`.